### PR TITLE
[Array-API] Add support for Box space

### PR DESCRIPTION
**Pull request by amacati**
# Description

This PR is a first shot at making spaces generic over all Array API compatible frameworks (e.g. `numpy`, `jax`, `torch`, ...).

The benefit of this change would be to make gymnasium compatible with the growing list of environments written in `jax` or `torch`. Strictly speaking, the `ArrayConversion` wrappers (e.g. `JaxToTorch`) are already outside of what we can currently express with `gymnasium.spaces`: There is no way to let the action and observation spaces reflect that we are expecting `jax.Array`s as observations and `torch.Tensor`s as inputs (safe for custom spaces which are rarely used).

This PR is also meant as a basis for discussion if things require breaking changes, new core dependencies (i.e. `array_api_compat` would become a core dependency) etc. 

One fundamental change is that each box is now also linked with a `device`. This allows users to express that e.g. the observation is an array on the GPU. The example below shows how this impacts `space.contains`.

## Open Challenges
- [ ] `space.sample`: Sampling will require several changes. It does not make sense to seed a `torch` space with a `numpy` random generator
- [ ] All tests can be rewritten to test against all Array API compatible frameworks. I would like to postpone this until we have a clearer picture of how things should be implemented though

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Example
With the changes in this PR, it is already possible to run the following:

```python
import array_api_strict as xp
import jax.numpy as jp
import numpy as np
import torch

from gymnasium.spaces import Box


box = Box(low=np.zeros(3, dtype=np.float32), high=np.zeros(3, dtype=np.float32))
print(box)  # Box(0.0, 0.0, (3,), <class 'numpy.float32'>, cpu)

strict_box = Box(low=xp.zeros(3, dtype=xp.float32), high=xp.zeros(3, dtype=xp.float32))
print(strict_box)  # Box(Array(0., dtype=array_api_strict.float32), Array(0., dtype=array_api_strict.float32), (3,), array_api_strict.float32, array_api_strict.Device('CPU_DEVICE'))

torch_box = Box(low=torch.zeros(3), high=torch.ones(3))
print(torch_box)  # Box(tensor(0.), tensor(1.), (3,), torch.float32, cpu)

jax_box = Box(low=jp.zeros(3), high=jp.ones(3))
print(jax_box)  # Box(0.0, 1.0, (3,), <class 'jax.numpy.float32'>, cuda:0)

cpu_device = jax.devices("cpu")[0]
print(jax_box.contains(jp.array([0.5, 0.5, 0.5], device=cpu_device)))  # False, devices do not match
gpu_device = jax.devices("gpu")[0]
print(jax_box.contains(jp.array([0.5, 0.5, 0.5], device=gpu_device)))  # True
```

# Checklist:

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
I'm starting to believe this is a sort of a Gymnasium 2.0 like feature change. 
This would give us more liberties for changing the API where necessary though I wish to make a few of that as possible. 
It would allow us to drop NumPy 1.x for 2.2+
Also we could simplify the Box space if we wished as currently it looks like a mound of technical debt to me.

Thoughts? If we doing this, it would be good to get some more voices involved to check their thoughts as well

**Comment by amacati** [comment]
I would definitely be on board with converting everything. The spaces were just a good module to start with. I also agree with the technical dept. Box could be simplified by a lot if `low` and `high` were always arrays.

This could very well be something for a gymnasium 2.0 release and that it would be great to get some more voices from other people to see where the different interests lie at.

**Comment by Jammf** [comment]
I agree that it'd make more sense as a 2.0 change, and a good opportunity to clean up the tech debt with `Box` with more dramatic changes. I also like the idea of having `low` and `high` be Arrays, constrained to having the same backing library, dtype, and device (and perhaps shape?). Then it'd be impossible to pass in values for `low`/`high` that are outside the bounds of the Box's `dtype`, so all that logic could be removed.

It's also make typing `Box` easier too, since then we could have something like:

```python
Box(Space[T]):
    def __init__(self, low: T, high: T, seed: SeedType): ...
    def contains(x: T) -> bool: ...
    def sample() -> T: ...
```

with `T` having the appropriate bounds. Then, if numpy (or other libraries) ever get something shape typing, that should just work without any changes on our end.

**Comment by Jammf** [comment]
If all the implementation is removed from it, I think making `Env` a Protocol makes sense. That might also be a good time to add a `RenderableEnv` subprotocol to make rendering support optional, along the lines of what #842 was proposing.

**Comment by amacati** [comment]
If we want this to be more of a 2.0 feature I could also draft a version with significantly reduced complexity which would not be backward compatible. But it does not make much sense to iterate more on this PR before we are sure what direction this should take
